[P116] Add alternative model selections for ST7789/ST7735

tonhuisman opened the pull request:
From a [Forum request](https://www.letscontrolit.com/forum/viewtopic.php?t=7229#p68480)

Features:
- Add extra device initialization sequences for ST7789 displays on devices like the LilyGO TTGO T-Display ~and T-Display S3~ (Excluded for now)
  - `(alt1)` -> LilyGO T-Display repository code
  - `(alt2)` -> Bodmer TFT_eSPI repository code (generic mode)
  - `(alt3)` -> Bodmer TFT_eSPI repository code (Sequence_3, intended for TTGO T-Display S3)
- Add extra device initialization sequences for ST7735 displays on devices like the LilyGO TTGO T-Display ~and T-Display S3~
  - Added support for resolution 135x240, using similar initialization like the 128x160 model, with corrections for resolution and rotation parameters

TODO:
- [x] Testing by requester ([confirmed](https://github.com/letscontrolit/ESPEasy/pull/5003#issuecomment-2020354121))
- [x] Documentation updates

tonhuisman committed:
[P116] Add alternative model selections for ST7789

plin2 commented:
flashed Y:\Temp\Bin-ESP32-max_ESP32_16M8M_LittleFS-20240309_PR_5003_8215939347\ESP_Easy_mega_20240309_**max**_ESP32_16M8M_LittleFS.factory.bin

configured hardware and device

accidentially TFT display model was set to ST7735 -> bright screen, little white square, rest looke like tv snow after end of broadcast

corrected TFT display model to ST7789 -> the usual dark grey screen, can't identofy text in line 1
corrected TFT display model to ST7789 (alt 1)-> gpio,4,1-> no effect, st77xx,txt,Hallo -> no effect
corrected TFT display model to ST7789 (alt 2)-> gpio,4,1-> no effect, st77xx,txt,Hallo -> no effect
corrected TFT display model to ST7789 (alt 3)-> gpio,4,1-> no effect, st77xx,txt,Hallo -> no effect

changed to TFT display ST7735 again, at least I get a white square, gpio,4,1-> has effect, st77xx,txz,64,64,XXX -> shows text

looks like the ST7735 is working pretty well, just that the geometrics of the display is not correct and there is an offset

[Update]

ran some structured tests (see attached images):

##############################################

ST7735 128 x 128 px
-------------------
st77xx,txc,black,white

st77xx,clear

st77xx,rf,1,1,10,10,red -> no-show
st77xx,rf,1,118,10,10,orange -> no-show
st77xx,rf,118,1,10,10,yellow -> no-show
st77xx,rf,118,118,10,10,green,red -> yellow

st77xx,r,54,54,10,10,red -> yellow
st77xx,r,118,54,10,10,orange -> dark yellow
st77xx,r,54,118,10,10,yellow -> red

##############################################

ST7735 80 x 160 px
-------------------
st77xx,txc,black,white

st77xx,clear

st77xx,rf,1,1,10,10,red -> no-show
st77xx,rf,1,118,10,10,orange -> no-show
st77xx,rf,70,1,10,10,yellow -> no-show
st77xx,rf,118,118,10,10,green,red -> yellow

st77xx,r,54,54,10,10,red -> turquoise
st77xx,r,34,118,10,10,orange -> dark turquoise
st77xx,r,54,118,10,10,yellow -> blue

![ST7735_128x128](https://github.com/letscontrolit/ESPEasy/assets/3195189/27ef62ea-ed5d-48ad-9b5c-a4b726b3b99d)
![ST7735_80x160](https://github.com/letscontrolit/ESPEasy/assets/3195189/9a8325ea-cbf3-4dcf-a0eb-f167ccd1d249)

----

flashed Y:\Temp\Bin-ESP32-display_ESP32_4M316k-20240309_PR_5003_8215939347\ESP_Easy_mega_20240309_**display**_ESP32_4M316k.factory.bin

no chnage vs. the max version. All ST7789 settings remain dead, ST7735 works better

tonhuisman commented:
Thanks for the feedback. I'll have to wait for my ordered unit to arrive before I can do any useful further investigation 🤔

plin2 commented:
No problem, I'll be patient (I hope).

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P116-add-alternative-st7789-revisions

tonhuisman committed:
[P116] Add alternative model selection for ST7735

tonhuisman commented:
@plin2 I've added support for another **ST7735** driver, with 135x240 pixel resolution and corrected rotation mapping (tested on the same hardware you use 😃).
This unit seems to have an ST7735 display controller, despite the manufacturer claiming it to be an ST7789. The initialization sequences (and chips...) are quite close, but it still wouldn't work with _any_ of the ST7789 drivers I tried.

NB: The ST7789 alt1..alt3 options are currently disabled in the plugin, as these haven't been verified on actual hardware.

Please test and report your findings 😺

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P116-add-alternative-st7789-revisions

plin2 commented:
Hi Ton,

missed the email with the update info. Just saw it. Where do I find the file for that version?

Thx, Peter

tonhuisman commented:
> Hi Ton,
> 
> missed the email with the update info. Just saw it. Where do I find the file for that version?
> 
> Thx, Peter

The All checks... link has the answer: https://github.com/letscontrolit/ESPEasy/actions/runs/8318298567?pr=5003
There you can download exactly the build you want to use, with the code from this PR 😄

plin2 commented:
The quick version of the test results:

st77xxcmd,off -> ok
st77xxcmd,on -> ok
st77xxcmd,clear -> ok
st77xxcmd,clear,black -> NOT ok, right column of pixels remains
st77xxcmd,clear,red -> NOT ok, no change
st77xx,clear,green -> ok
st77xxcmd,clear,black -> ok, but right column of pixels remains

st77xx,l,1,1,135,240,red -> ok
st77xx,lv,10,120,green -> ok
st77xx,lv,20,20,blue -> ok	
st77xx,r,1,1,10,20,white -> ok
st77xx,r,1,1,135,240,purple ->  Command unknown
st77xx,r,1,1,134,239,purple ->  ok, see image

Main problem: At rotation 0 there is a problem with text output. The first pixels in a row seem to appear in the last row of pixels. 
![Rotation0_Text1](https://github.com/letscontrolit/ESPEasy/assets/3195189/750ea55e-8407-473c-ad76-958f60ad7f47)
![Rotation0_Text2](https://github.com/letscontrolit/ESPEasy/assets/3195189/baab1016-27d4-4e04-b683-cabd1779faa3)
![Rotation0_Text3](https://github.com/letscontrolit/ESPEasy/assets/3195189/76d2a7a3-aa32-4eb5-b639-17c21b4fcbcf)
![Rotation90_Text](https://github.com/letscontrolit/ESPEasy/assets/3195189/737c6058-209e-4bde-8557-fcafb37ab89d)
![Rotation180_Text](https://github.com/letscontrolit/ESPEasy/assets/3195189/530b3434-7c54-4346-9aae-4d5c32d69114)
![Rotation0_Drawing](https://github.com/letscontrolit/ESPEasy/assets/3195189/96f93978-3c65-4d42-8a3a-130c30846256)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/P116-add-alternative-st7789-revisions

tonhuisman commented:
> st77xxcmd,clear,black -> NOT ok, right column of pixels remains
> st77xxcmd,clear,red -> NOT ok, no change

Hm, this is a ~bug~ unclarity, though actually in the documentation. the `<trigger>cmd,clear` command _doesn't_ support providing a color to clear the background, though the `<trigger>,clear[,<color>]` command _does_ support an optional color. If no color is provided (or supported), the default background color is used (I've set mine to `navy`, just to test that, and because I like it as the background color 😛)
Edit: The documentation wasn't wrong, but could be more explicit, so that's now improved.

The garbage pixels are fixed in the next commit.

> st77xx,r,1,1,135,240,purple -> Command unknown

Well, technically, that's correctly refused. A check is done if the rectangle can be drawn within the confines of the display (or defined window, but that's a different story, see the docs for the `defwin` sub-command family 😉). The 135 and 240 arguments are the **width** and **height** of the rectangle, and as you are starting at position 1,1, it would draw the right-top at position 1,136, and that's outside the available 135/240 window.
Window coordinates start at position 0,0, but as the offset isn't correct yet (fixed in the next commit), you'd see a partially drawn rectangle.

tonhuisman committed:
[P116] Fine-tuning the rotation column-offset for ST7735 135x240 display

tonhuisman committed:
[P116] Documentation improvement.

tonhuisman committed:
[P095] Documentation improvement.

tonhuisman commented:
@plin2 You can test again when this Actions run is completed.

Btw: You're using the SPIFFS MAX 16M1M build, but that will stop to exist sometime in the near future, as the deprecated SPIFFS filesystem (already removed from IDF 5.1, but we're using a specially crafted edition that temporarily still includes SPIFFS) will be replaced by LittleFS. An automatic, on-board, migration from SPIFFS to LittleFS is not possible, so that'll involve manual intervention.

You are **strongly** advised to use one of the MAX 16M8M LittleFS builds, to a) avoid future SPIFFS -> LittleFS migration, and b) have a bigger filesystem for storing f.e. images to show on the display 😉

plin2 commented:
Here arte some fresh test resultzs:

st77xxcmd,off -> ok
st77xxcmd,on -> ok
st77xxcmd,clear -> but clears to last background color
st77xxcmd,clear,black -> no change
st77xx,clear,black -> ok, clears to black background
st77xxcmd,clear,red -> but clears to last background color
st77xx,clear,red -> ok, clears to red background color
st77xx,clear,green -> ok

st77xx,rot,3     90 degrees
st77xx,rot,0    180 degrees
st77xx,rot,2      0 degrees

with rotation set to +180°
st77xx,clear,black
st77xx,l,0,0,135,240,red -> ok
st77xx,lv,10,120,green -> ok
st77xx,lv,20,20,blue -> ok	
st77xx,r,0,0,10,20,white -> ok
st77xx,r,0,0,135,240,purple -> ok 

with rotation set back to +0°
st77xx,clear,black
st77xx,r,0,0,240,135,purple -> ok 

Funnily st77xx**cmd**,clear,black doesn't clear to the given color, but st77xx,clear,black does.

plin2 commented:
> @plin2 You can test again when this Actions run is completed.
> You are **strongly** advised to use one of the MAX 16M8M LittleFS builds, to a) avoid future SPIFFS -> LittleFS migration, and b) have a bigger filesystem for storing f.e. images to show on the display 😉

ok, I'll switch to that now

tonhuisman commented:
> > @plin2 You can test again when this Actions run is completed.
> > You are **strongly** advised to use one of the MAX 16M8M LittleFS builds, to a) avoid future SPIFFS -> LittleFS migration, and b) have a bigger filesystem for storing f.e. images to show on the display 😉
> 
> ok, I'll switch to that now

You either have to save the configuration beforehand (Tools/Save), and restore (Tools/Load) (after renaming to `config.dat`) when uploading the different flash-layout build. Or reconfigure the device 😃 



> Funnily st77xx**cmd**,clear,black doesn't clear to the given color, but st77xx,clear,black does.

As said, that's the documented behavior 😉, `<trigger>cmd` is handled by the plugin, and the `<trigger>` behavior is handled by the generic ESPEasy AdafruitGFX Helper module, as used by the Adafruit GFX based display plugins (P095, P096, P116, P131 and P141).

plin2 commented:
From my point of view it's working now, unless you have other tests that I can run. 

Thx for the support
Peter

ironataerial commented:
>  rest looke like tv snow after end of broadcast

Tell me how old you are without telling me how old you are :)

plin2 commented:
> > rest looke like tv snow after end of broadcast
> 
> Tell me how old you are without telling me how old you are :)

Actually, once the display worked I wantetd to reply: 
When I got my Tandy TRS80 with a BASIC manual the first steps where: type 
10 Print "Hello"
Then type "run" and shout "**Ma, it Works!**"

tonhuisman committed:
[P116] Update documentation for supported displays
